### PR TITLE
more aggressive NN early stopping

### PR DESCRIPTION
**Pull request by jwmueller**
Change NN to early-stop if validation-score hasn't *strictly* improved in `epochs_wo_improve` consecutive epochs.
Suppose epoch `e1` = epoch where validation-score first hits max-value and validation-score plateaus at this point. 
After this change, we now consider the "best" model the one from epoch `e2 = e1 + epochs_wo_improve`. 
Thus we still train a little bit beyond the point where validation-score plateaus (perhaps margin of decision-boundary can still be improved), but the amount of extra-training is strictly limited by the user-controllable hyperparameter `epochs_wo_improve` (rather than before where the NN could continue training for a long time while validation-score remained plateaued).

So users who want to be sure they are achieving maximal accuracy (at potential cost of long training-time) can still do this by appropriately increasing the `epochs_wo_improve` hyperparameter. 

*Issue #, if available:*

https://github.com/awslabs/autogluon/issues/595

### CODE REVIEW

**Comment by jwmueller** [comment]
I believe this strategy is preferred vs the simple approach of just returning the model from epoch `e1` (defined above).
And does not involve extra computation as we anyway need to train the extra `epochs_wo_improve` epochs to determine that the validation-loss is no longer improving.

**Comment by Innixma** [comment]
> @Innixma I believe this strategy is preferred vs the simple approach of just returning the model from epoch `e1` (defined above).
> And does not involve extra computation as we anyway need to train the extra `epochs_wo_improve` epochs to determine that the validation-loss is no longer improving.

Say `epochs_wo_improve=5`

```
E0: 0.5 auc
E1: 0.6 auc
E2: 0.7 auc
E3: 0.8 auc
E4: 0.8 auc
E5: 0.8 auc
E6: 0.7 auc
E7: 0.6 auc
E8: 0.6 auc
```

In this case with the proposed logic, `e1=3`, and `e2=8`

Selecting `e2=8` is not ideal. e2 should either be 3, 4, or 5. I propose 5, such that we keep track of the last epoch which matched the validation score of `e1`, and select that epoch when early stopping.

**Comment by szha** [comment]
Job PR-598-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-598/1/index.html

**Comment by szha** [comment]
Job PR-598-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-598/2/index.html

**Comment by jwmueller** [comment]
> If you still say that `e2=8` is optimal, are you saying that we should not trust the validation score, and that we intrinsically trust that the model is improving?

No sorry, my example was only referring to the case where validation score plateaus (is totally constant for all epochs past `e1`).
If your example were instead modified to have: val-score = 0.8 for all epochs >= 3, then yes I'm suggesting `e2=8` is the one we want to return with `epochs_wo_improve=5`.

In the case of your actual example, the code in this PR should return: `e2= 5`, which I do think is better than `e2=3` for reasons of potential margin-increase.  To translate your example into the code of the PR, basically `best_val_epoch = 5` but  `val_improve_epoch` would remain stuck at 3, so training would still `break` at E8 with `epochs_wo_improve=5`.

**Comment by Innixma** [comment]
> @Innixma
> 
> > If you still say that `e2=8` is optimal, are you saying that we should not trust the validation score, and that we intrinsically trust that the model is improving?
> 
> No sorry, my example was only referring to the case where validation score plateaus (is totally constant for all epochs past `e1`).
> If your example were instead modified to have: val-score = 0.8 for all epochs >= 3, then yes I'm suggesting `e2=8` is the one we want to return with `epochs_wo_improve=5`.
> 
> In the case of your actual example, the code in this PR should return: `e2= 5`, which I do think is better than `e2=3` for reasons of potential margin-increase. To translate your example into the code of the PR, basically `best_val_epoch = 5` (and this is the model that's saved/reloaded for prediction) but `val_improve_epoch` would remain stuck at 3, so training would still `break` at E8 with `epochs_wo_improve=5`.

I see, makes sense. This differs from what your original description was so I originally thought it was doing `e2=8` because of `e2 = e1 + epochs_wo_improve`, which is not quite accurate to the code.

I think this code is correct, but before I approve I will benchmark before/after versions to ensure we don't experience degradation of performance.

**Comment by szha** [comment]
Job PR-598-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-598/3/index.html

**Comment by szha** [comment]
Job PR-598-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-598/4/index.html

**Comment by szha** [comment]
Job PR-598-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-598/5/index.html

**Comment by szha** [comment]
Job PR-598-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-598/6/index.html

**Comment by Innixma** [comment]
Did a benchmark of this change. Looks to be mostly matching previous scores with 1 major change in the `kr-vs-kp` dataset.

`Prior score: 0.999961 AUC`
`New score: 0.999883 AUC`

Average epochs trained by NN went from 300 -> 27
The NN is the best model here, and it gets a very very high validation score (often 1). So after the change, the model likely wasn't fully fit even though it had a perfect validation AUC score, which results in a negligible (absolute) and major (relative) error increase.

On the upside, the NN training took only 1/10th of the time for `kr-vs-kp`, and the NN was previously taking 90%+ of the total training time among all models.

I think this will only ever occur on extremely easy problems, where perhaps the difference really isn't that important, and it could be argued we shouldn't be benchmarking on such easy datasets to begin with.

**Comment by Innixma** [comment]
Here is the run for creditfraud on my laptop to show that it early stops reasonably:

```
Epoch 0.  Train loss: 0.06224216, Val accuracy: 0.9996489996489997
Epoch 1.  Train loss: 0.0044396217, Val accuracy: 0.9996489996489997
Epoch 2.  Train loss: 0.0030911337, Val accuracy: 1.0
Epoch 3.  Train loss: 0.0026924242, Val accuracy: 0.9996489996489997
Epoch 4.  Train loss: 0.0022474146, Val accuracy: 0.9996489996489997
Epoch 5.  Train loss: 0.0021056526, Val accuracy: 0.9992979992979993
Epoch 6.  Train loss: 0.0018715413, Val accuracy: 1.0
Epoch 7.  Train loss: 0.0017852911, Val accuracy: 0.9996489996489997
Epoch 8.  Train loss: 0.0015773426, Val accuracy: 1.0
Epoch 9.  Train loss: 0.0014420052, Val accuracy: 0.9996489996489997
Epoch 10.  Train loss: 0.0013045863, Val accuracy: 1.0
Epoch 11.  Train loss: 0.0011904336, Val accuracy: 1.0
Epoch 12.  Train loss: 0.0011358258, Val accuracy: 1.0
Epoch 13.  Train loss: 0.0010043315, Val accuracy: 1.0
Epoch 14.  Train loss: 0.00096173154, Val accuracy: 1.0
Epoch 15.  Train loss: 0.0009741376, Val accuracy: 1.0
Epoch 16.  Train loss: 0.00090411986, Val accuracy: 1.0
Epoch 17.  Train loss: 0.00074588344, Val accuracy: 1.0
Epoch 18.  Train loss: 0.0008922371, Val accuracy: 0.9996489996489997
Epoch 19.  Train loss: 0.0007041674, Val accuracy: 1.0
Epoch 20.  Train loss: 0.00066614343, Val accuracy: 1.0
Epoch 21.  Train loss: 0.00063342834, Val accuracy: 1.0
Epoch 22.  Train loss: 0.00059692626, Val accuracy: 1.0
Epoch 23.  Train loss: 0.0006864279, Val accuracy: 1.0
Best model found in epoch 23. Val accuracy: 1.0
Saving AutogluonModels/ag-20200828_021941/models/NeuralNetClassifier/model.pkl
	1.0	 = Validation accuracy score
	808.1s	 = Training runtime
	0.28s	 = Validation runtime
```

**Comment by jwmueller** [comment]
> I think this will only ever occur on extremely easy problems, where perhaps the difference really isn't that important, and it could be argued we shouldn't be benchmarking on such easy datasets to begin with.

Right this is what I've been saying all along regarding the NN continuing to expand the margin (for datasets where classes are entirely separable so 100% test accuracy is possible).  So there'll always be some tension between runtime and test accuracy for certain datasets, but at least now the user can adjust it themselves via `epochs_wo_improve` hyperparameter. We may consider different default value of this hyperparameter going forward as well, the current choice of 20 is fairly arbitrary...
